File: src/main/java/com/founderrd/xda/XDAByteArrayInputStream.java
/**
 * Title:	XDAByteArrayInputStream
 * Description:	以ByteArrayInputStream为底层实现XDAInputStream
 * Author:	杨天航(tianhang.yang@gmail.com)
 * version: 	1.0
 * time: 	2009.11.11
 */

package com.founderrd.xda;

import javax.annotation.Nonnull;
import java.io.ByteArrayInputStream;
import java.io.IOException;

class XDAByteArrayInputStream extends XDAInputStream {
    private final byte[] buffer;
    private ByteArrayInputStream buffStream;

    XDAByteArrayInputStream(byte[] buffer) {
        this.buffer = buffer;
        buffStream = null;
    }

    public int available() throws IOException {
        return buffStream.available();
    }

    public void close() throws IOException {
        buffStream.close();
    }

    public void mark(int readlimit) {
        buffStream.mark(readlimit);
    }

    public boolean markSupported() {
        return buffStream.markSupported();
    }

    public int read() throws IOException {
        return buffStream.read();
    }

    public int read(@Nonnull byte[] b) throws IOException {
        return buffStream.read(b);
    }

    public int read(@Nonnull byte[] b, int off, int len) throws IOException {
        return buffStream.read(b, off, len);
    }

    public void reset() throws IOException {
        buffStream.reset();
    }

    public long skip(long n) throws IOException {
        return buffStream.skip(n);
    }

    public boolean open() throws IOException {
        if (buffStream != null)
            return false;
        buffStream = new ByteArrayInputStream(this.buffer);
        return true;
    }

    public XDAInputStream nirvana() throws IOException {
        if (buffStream != null)
            buffStream.close();

        return new XDAByteArrayInputStream(this.buffer);
    }

    public boolean canRead() {
        return (buffStream != null);
    }
}

File: src/main/kotlin/com/founderrd/xda/XDAByteArrayInputStream.kt
/**
 * Title:	XDAByteArrayInputStream
 * Description:	以ByteArrayInputStream为底层实现XDAInputStream
 * Author:	杨天航(tianhang.yang@gmail.com)
 * version: 	1.0
 * time: 	2009.11.11
 */
package com.founderrd.xda

import java.io.ByteArrayInputStream
import java.io.IOException

internal class XDAByteArrayInputStream(private val buffer: ByteArray) : XDAInputStream() {

    private var buffStream: ByteArrayInputStream? = null

    @Throws(IOException::class)
    override fun available(): Int {
        return buffStream!!.available()
    }

    @Throws(IOException::class)
    override fun close() {
        buffStream!!.close()
    }

    override fun mark(readlimit: Int) {
        buffStream!!.mark(readlimit)
    }

    override fun markSupported(): Boolean {
        return buffStream!!.markSupported()
    }

    @Throws(IOException::class)
    override fun read(): Int {
        return buffStream!!.read()
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        return buffStream!!.read(b)
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        return buffStream!!.read(b, off, len)
    }

    @Throws(IOException::class)
    override fun reset() {
        buffStream!!.reset()
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        return buffStream!!.skip(n)
    }

    @Throws(IOException::class)
    override fun open(): Boolean {
        if (buffStream != null) return false
        buffStream = ByteArrayInputStream(buffer)
        return true
    }

    @Throws(IOException::class)
    override fun nirvana(): XDAInputStream? {
        if (buffStream != null) buffStream!!.close()
        return XDAByteArrayInputStream(buffer)
    }

    override fun canRead(): Boolean {
        return buffStream != null
    }
}